Actively evict content in ManagedSourceBuffer under memory pressure.

jyavenard opened the pull request:
#### e994c1218d752469951b880e4c2ca6021cfc79f9
<pre>
Actively evict content in ManagedSourceBuffer under memory pressure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=252735">https://bugs.webkit.org/show_bug.cgi?id=252735</a>
rdar://105769723

Reviewed by Jer Noble.

We call the Coded Frame Eviction algorithm when the content process receives
a memory pressure signal.
The eviction algorithm used is identical as the non-managed SourceBuffer&apos;s one.
Where the SourceBuffer will stop the eviction process as soon as sufficient
memory has been freed, with a ManagedSourceBuffer we evict as much as we
can over the entire source buffered range.
If an eviction occurred, the bufferedchange event will be fired on the
ManagedSourceBuffer.

* LayoutTests/platform/glib/TestExpectations: Test is failing as no WebProcess memory pressure handler is installed.
* LayoutTests/media/media-source/media-managedmse-memorypressure-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-memorypressure.html: Added.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::memoryPressure):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::memoryPressure):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setBufferingPolicy):
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::operator== const):
(WebCore::PlatformTimeRanges::operator!= const):
* Source/WebCore/platform/graphics/PlatformTimeRanges.h: Add operator== and operator!= support.
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::evictCodedFrames):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::evictFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::memoryPressure):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::memoryPressure):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/260825@main">https://commits.webkit.org/260825@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf4e246371a869199f1f323737fd38768dd45f5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18700 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1065 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113461 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9897 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101846 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115333 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98242 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29895 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11436 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12094 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17459 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50846 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7504 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13834 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+
+bool SourceBufferPrivate::evictFrames(uint64_t newDataSize, uint64_t maximumBufferSize, const MediaTime& currentTime, bool isEnded)
+{
+    bool isBufferFull = true;
```
Nit: auto

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+    // FIXME: All this is nice but we should take into account negative playback rate and begin from after current time
+    // and be more conservative with before current time.
+
+    unsigned timeChunkAsMilliseconds = evictionAlgorithmInitialTimeChunk;
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+    unsigned timeChunkAsMilliseconds = evictionAlgorithmInitialTimeChunk;
+    do {
+        const MediaTime timeChunk = MediaTime(timeChunkAsMilliseconds, 1000);
+        const MediaTime maximumRangeEnd = std::min(currentTime - timeChunk, findPreviousSyncSamplePresentationTime(currentTime));
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+
+        do {
+            MediaTime rangeStart = buffered.minimumBufferedTime();
+            MediaTime rangeEnd = std::min(rangeStart + timeChunk, maximumRangeEnd);
```
Ditto

jyavenard replied:
I'm really not a fan on the use of `auto` on return value that don't really tell you what the return type is.

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+            removeCodedFrames(rangeStart, rangeEnd, currentTime, isEnded);
+            MediaTime newRangeStart = buffered.minimumBufferedTime();
+            if (newRangeStart == rangeStart)
+                break;
```
Nit: don't really need this as a local variable.

jyavenard replied:
that's existing code I moved; though I can see the value during development/debugging. Easier to step in the debugger and look at the value of the variable after a breakpoint

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+    } while (isBufferFull && timeChunkAsMilliseconds >= evictionAlgorithmTimeChunkLowThreshold);
+
+    if (!isBufferFull)
+        return isBufferFull;
```
`return true` is more obvious

eric-carlson commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -1103,6 +1034,87 @@ void SourceBufferPrivate::append(Vector<unsigned char>&&)
+            // 4. For each range in removal ranges, run the coded frame removal algorithm with start and
+            // end equal to the removal range start and end timestamp respectively.
+            removeCodedFrames(rangeStart, rangeEnd, currentTime, isEnded);
+            MediaTime newRangeEnd = buffered.maximumBufferedTime();
```
`auto`

jyavenard committed:
Actively evict content in ManagedSourceBuffer under memory pressure.
https://bugs.webkit.org/show_bug.cgi?id=252735
rdar://105769723

Reviewed by Jer Noble.

We call the Coded Frame Eviction algorithm when the content process receives
a memory pressure signal.
The eviction algorithm used is identical as the non-managed SourceBuffer's one.
Where the SourceBuffer will stop the eviction process as soon as sufficient
memory has been freed, with a ManagedSourceBuffer we evict as much as we
can over the entire source buffered range.
If an eviction occurred, the bufferedchange event will be fired on the
ManagedSourceBuffer.

* LayoutTests/platform/glib/TestExpectations: Test is failing as no WebProcess memory pressure handler is installed.
* LayoutTests/media/media-source/media-managedmse-memorypressure-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-memorypressure.html: Added.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::memoryPressure):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::memoryPressure):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setBufferingPolicy):
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::operator== const):
(WebCore::PlatformTimeRanges::operator!= const):
* Source/WebCore/platform/graphics/PlatformTimeRanges.h: Add operator== and operator!= support.
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::evictCodedFrames):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::evictFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::memoryPressure):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::memoryPressure):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: https://commits.webkit.org/260825@main
